src/main/daml/Daml/Finance/Settlement/Instruction.daml

```diff
@@ -119,11 +121,13 @@ template Instruction
             discloseA receiverAccount
             vReceiverAccount <- view <$> fetchInterfaceByKey @Account.R receiverAccount
             assertMsg "actors must include approvers" $ S.isSubsetOf vReceiverAccount.controllers.approvers actors
+            assertMsg "the custodian of the receiver account does not match the target custodian" $ step.custodian == receiverAccount.custodian
           PassthroughTo (passthroughAccount, toInstruction) -> do
             discloseA passthroughAccount
             vReceiverAccount <- view <$> fetchInterfaceByKey @Account.R passthroughAccount
             assertMsg "actors must include approvers" $ S.isSubsetOf vReceiverAccount.controllers.approvers actors
           _ ->
+            -- we could add a check receiver == targetCustodian in the DebitSender case (not the off-ledger one). Not sure it is needed given that we can't Debit the holding if custodian does not authorize
```

Comment:
I think it makes sense to add the check, as it'll give us a better error message. Same above with `CreditReceiver`.